veg2hab/mozaiek.py

```diff
@@ -93,76 +104,301 @@ def __str__(self):
 
-    def __str__(self):
-        complete_string = ""
-        complete_string += f"{'(als rand langs)' if self.ook_als_rand_langs else ''} "
-        complete_string += (
-            f"{'zelfstndg' if self.alleen_zelfstandig else 'zelfstndg/mozk'} "
+    def _vul_mozk_perc_dict(self, omringd_door: pd.DataFrame) -> None:
+        """
+        Vult de mozk_perc_dict met keys (habtype, kwaliteit) en values (omringing percentage)
+
+        Is gescheiden van _bepaal_kwalificerende_en_HXXXX_omringing om de logica netter te houden
+        """
+        mozk_perc_dict = defaultdict(int)
+
+        grouped_by_vlak = omringd_door.groupby("ElmID")
+
+        for _, vlak_group in grouped_by_vlak:
+            assert (
+                len(vlak_group.omringing_percentage.unique()) == 1
+            ), f"Omringing percentage is niet hetzelfde voor alle complexdelen in vlak met ElmID {vlak_group.ElmID.iloc[0]}"
+
+            grouped_by_habtype = vlak_group.groupby("habtype")
+            for habtype, habtype_group in grouped_by_habtype:
+                bedekking_goed_habtype = habtype_group[
+                    habtype_group.kwaliteit == Kwaliteit.GOED
+                ].complexdeel_percentage.sum()
+
+                if bedekking_goed_habtype >= self.mozaiek_minimum_bedekking:
+                    mozk_perc_dict[(habtype, Kwaliteit.GOED)] += habtype_group.iloc[
+                        0
+                    ].omringing_percentage
+                    continue
+
+                bedekking_matig_habtype = habtype_group.complexdeel_percentage.sum()
+                if bedekking_matig_habtype >= self.mozaiek_minimum_bedekking:
+                    if habtype in ["H0000", "HXXXX"]:
+                        # H0000 en HXXXX hebben altijd kwaliteit NVT
+                        mozk_perc_dict[(habtype, Kwaliteit.NVT)] += habtype_group.iloc[
+                            0
+                        ].omringing_percentage
+                    else:
+                        mozk_perc_dict[
+                            (habtype, Kwaliteit.MATIG)
+                        ] += habtype_group.iloc[0].omringing_percentage
+
+        # We slaan dit op als tuples ipv gewoon als de dict zodat we mozaiekregels kunnen serializen
+        self.mozk_perc_tuples = []
+
+        for (habtype, kwaliteit), perc in mozk_perc_dict.items():
+            self.mozk_perc_tuples.append(
+                MozkPercTuple(habtype=habtype, kwaliteit=kwaliteit, percentage=perc)
+            )
+
+    def _vul_tegengekomen_kwal_SBB_VvN(
+        self, tegengekomen_vegtypen: Set[Union[SBB, VvN]]
+    ):
+        # Casten naar list voor (de)serializatie
+        self.tegengekomen_kwal_SBB = list(
+            {vegtype for vegtype in tegengekomen_vegtypen if isinstance(vegtype, SBB)}
         )
+        self.tegengekomen_kwal_VvN = list(
+            {vegtype for vegtype in tegengekomen_vegtypen if isinstance(vegtype, VvN)}
+        )
+
+    def _bepaal_kwalificerende_en_HXXXX_omringing(
```

Comment:
docstring hier kan geen kwaad, code is moeilijk te begrijpen.

Reply:
Het is idd nogal gegoochel met percentages, thresholds en truth values. Hopelijk is het nu duidelijker, zo niet hoor ik t wel.